### PR TITLE
Issue 16042 - Identifier on template arguments should consider eponymous member lookup

### PR DESCRIPTION
**Pull request by 9rnsr**

### CODE REVIEW

**Comment by UplinkCoder** [comment]
LGTM.

File: src/expression.d [hunk 1]
```diff
@@ -1210,11 +1208,14 @@ extern (C++) bool arrayExpressionToCommonType(Scope* sc, Expressions* exps, Type
  */
 extern (C++) TemplateDeclaration getFuncTemplateDecl(Dsymbol s)
 {
-    FuncDeclaration f = s.isFuncDeclaration();
+    auto f = s.isFuncDeclaration();
     if (f && f.parent)
     {
-        TemplateInstance ti = f.parent.isTemplateInstance();
-        if (ti && !ti.isTemplateMixin() && ti.tempdecl && (cast(TemplateDeclaration)ti.tempdecl).onemember && ti.tempdecl.ident == f.ident)
+        auto ti = f.parent.isTemplateInstance();
```
Comment (UplinkCoder) [inline src/expression.d]: ditto

File: src/expression.d [hunk 2]
```diff
@@ -1210,11 +1208,14 @@ extern (C++) bool arrayExpressionToCommonType(Scope* sc, Expressions* exps, Type
  */
 extern (C++) TemplateDeclaration getFuncTemplateDecl(Dsymbol s)
 {
-    FuncDeclaration f = s.isFuncDeclaration();
+    auto f = s.isFuncDeclaration();
```
Comment (UplinkCoder) [inline src/expression.d]: please leave the typename.

Comment (9rnsr) [inline src/expression.d]: When we use downcast method, the target type name is already in rhs - is"FuncDeclaration". So this change happily reduces redundancy.

Comment (UplinkCoder) [inline src/expression.d]: ah that is what it does :)
isFuncDeclaration should really be renamed.

**Comment by georgi-d** [comment]
Is there a blocking issue for merging this? I would really like to have this resolved since I am constantly hitting this issue when dealing with complicated chains of algorithms.

Thanks

**Comment by WalterBright** [comment]
Took me a while to figure out that all the changed in expression.d are refactorings and have nothing to do with 16042. Please do not combine things like this.

**Comment by wilzbach** [comment]
@georgi-d 

> Is there a blocking issue for merging this? I would really like to have this resolved since I am constantly hitting this issue when dealing with complicated chains of algorithms.
> Thanks

Well @9rnsr is currently not active. This issue is blocked because the PR mixed refactoring / cleanup in `expression.d`.
=> The quickest way to get this into DMD is to take the PR from @9rnsr - split it in two pieces (bug fix & cleanup) and resubmit both. Sounds silly, but works ;-)

**Comment by WalterBright** [comment]
Rather than working on this enhancement, I'd prefer working on Kenji's regression fixes.

**Comment by JinShil** [comment]
Revived at #7404

**Pull request closed without merging**